PWGCF/TwoParticleCorrelations/Tasks/r2p2-4-id.cxx

```diff
@@ -0,0 +1,1175 @@
+        histos.fill(HIST("h2d_pt_etaPhiP"), track1.eta(), track1.phi(), track1.pt());
+      } else { //-ve particles
+        histos.fill(HIST("h2d_n1_etaPhiM"), track1.eta(), track1.phi());
+        histos.fill(HIST("h2d_pt_etaPhiM"), track1.eta(), track1.phi(), track1.pt());
+        histos.fill(HIST("h1d_n1_ptM"), track1.pt(), 1.0 / (2.0 * constants::math::PI * track1.pt()));
+      }
+      etabin1 = (track1.eta() + 0.8) * 15;
+      phibin1 = 36 * track1.phi() / (2 * constants::math::PI);
+      mult++;
+      //------------------------------------------------------------------------
+      for (auto track2 : tracks) {
+        //-----Two Particle Distribution------------------------------------------
+        if (track1.index() == track2.index())
+          continue; // No auto-correlations
+
+        etabin2 = (track2.eta() + 0.8) * 15;
+        phibin2 = 36 * track2.phi() / (2 * constants::math::PI);
+        if ((etabin1 >= 0) && (etabin2 >= 0) && (phibin1 >= 0) && (phibin2 >= 0) && (etabin1 < 24) && (etabin2 < 24) && (phibin1 < 36) && (phibin2 < 36)) {
+          if (track1.sign() * track2.sign() == -1) // Unlike-Sign particles
+          {
+            histos.fill(HIST("h2d_n2_eta1Phi1Eta2Phi2PM"), 36 * etabin1 + phibin1 + 0.5, 36 * etabin2 + phibin2 + 0.5);
+            histos.fill(HIST("h2d_npt_eta1Phi1Eta2Phi2PM"), 36 * etabin1 + phibin1 + 0.5, 36 * etabin2 + phibin2 + 0.5, track2.pt());
+            histos.fill(HIST("h2d_ptn_eta1Phi1Eta2Phi2PM"), 36 * etabin1 + phibin1 + 0.5, 36 * etabin2 + phibin2 + 0.5, track1.pt());
+            histos.fill(HIST("h2d_ptpt_eta1Phi1Eta2Phi2PM"), 36 * etabin1 + phibin1 + 0.5, 36 * etabin2 + phibin2 + 0.5, track1.pt() * track2.pt());
+          } else if (track1.sign() == 1) { //+ve particle pair
+            histos.fill(HIST("h2d_n2_eta1Phi1Eta2Phi2PP"), 36 * etabin1 + phibin1 + 0.5, 36 * etabin2 + phibin2 + 0.5);
+            histos.fill(HIST("h2d_npt_eta1Phi1Eta2Phi2PP"), 36 * etabin1 + phibin1 + 0.5, 36 * etabin2 + phibin2 + 0.5, track2.pt());
+            histos.fill(HIST("h2d_ptn_eta1Phi1Eta2Phi2PP"), 36 * etabin1 + phibin1 + 0.5, 36 * etabin2 + phibin2 + 0.5, track1.pt());
+            histos.fill(HIST("h2d_ptpt_eta1Phi1Eta2Phi2PP"), 36 * etabin1 + phibin1 + 0.5, 36 * etabin2 + phibin2 + 0.5, track1.pt() * track2.pt());
+          } else { //-ve particle pair
+            histos.fill(HIST("h2d_n2_eta1Phi1Eta2Phi2MM"), 36 * etabin1 + phibin1 + 0.5, 36 * etabin2 + phibin2 + 0.5);
+            histos.fill(HIST("h2d_npt_eta1Phi1Eta2Phi2MM"), 36 * etabin1 + phibin1 + 0.5, 36 * etabin2 + phibin2 + 0.5, track2.pt());
+            histos.fill(HIST("h2d_ptn_eta1Phi1Eta2Phi2MM"), 36 * etabin1 + phibin1 + 0.5, 36 * etabin2 + phibin2 + 0.5, track1.pt());
+            histos.fill(HIST("h2d_ptpt_eta1Phi1Eta2Phi2MM"), 36 * etabin1 + phibin1 + 0.5, 36 * etabin2 + phibin2 + 0.5, track1.pt() * track2.pt());
+          }
+        }
```

Comment:
This is extremely inefficient!! The task will be hardly penalized regarding CPU usage!
The checks on the bin acceptability should not be done within the double inner tracks loop
The cascaded `if`s also impose a huge execution penalty
I would suggest caching the histograms pointers in indexed variables according to the kind of track. In this routine there is only `P` and `M` then you can have a 2D 2x2 matrix where the (0,0), (0,1), (1,0), and (1,1) correspond to the MM, MP, PM, and PP combinations. In a previous loop over the tracks you check the proper binning and assign 0 to the track if it is negative an 1 if it is positive, -1 if it is out of proper binning. Then in the loop just use the assigned indexes for addressing the proper histogram checking only that the indexes are not negative

Reply:
Sir, I don't know a way to assign the index to a track, other than using a vector to store the indexes. Is there any other way to do it?